tests/test_preprocess_pipeline.py

```diff
@@ -0,0 +1,115 @@
+from aspire.volume import Volume
+
+DATA_DIR = os.path.join(os.path.dirname(__file__), "saved_test_data")
+
+
+class PreprocessPLTestCase(TestCase):
+    def setUp(self):
+
+        self.L = 64
+        self.n = 128
+
+        self.noise_variance = 0.004
+        self.noise_filter = ScalarFilter(dim=2, value=self.noise_variance)
+
+        self.sim = Simulation(
+            L=self.L,
+            n=self.n,
+            unique_filters=[
+                RadialCTFFilter(defocus=d) for d in np.linspace(1.5e4, 2.5e4, 7)
+            ],
+            noise_filter=self.noise_filter,
+        )
+        self.imgs_org = self.sim.images(start=0, num=self.n)
+
+    def testPhaseFlip(self):
+        self.sim.phase_flip()
+        imgs_pf = self.sim.images(start=0, num=self.n)
+
+        # check energy conservation
+        self.assertTrue(np.allclose(self.imgs_org.norm(), imgs_pf.norm()))
+
+    def testDownsample(self):
+        # generate a 3D map with density decays as Gaussian function
+        g3d = grid_3d(self.L)
+        coords = np.array([g3d["x"].flatten(), g3d["y"].flatten(), g3d["z"].flatten()])
+        sigma = 0.2
+        vol = np.exp(-0.5 * np.sum(np.abs(coords / sigma) ** 2, axis=0))
+        vol = np.reshape(vol, g3d["x"].shape)
+        vols = Volume(vol)
+
+        # set noise to zero and CFT filters to unity for simulation object
+        noise_var = 0
+        noise_filter = ScalarFilter(dim=2, value=noise_var)
+        sim = Simulation(
+            L=self.L,
+            n=self.n,
+            vols=vols,
+            offsets=0.0,
+            amplitudes=1.0,
+            unique_filters=[
+                ScalarFilter(dim=2, value=1) for d in np.linspace(1.5e4, 2.5e4, 7)
+            ],
+            noise_filter=noise_filter,
+        )
+        # get images before downsample
+        imgs_org = sim.images(start=0, num=self.n)
+        # get images after downsample
+        max_resolution = 32
+        sim.downsample(max_resolution)
+        imgs_ds = sim.images(start=0, num=self.n)
+
+        # check resolution
+        self.assertTrue(np.allclose(max_resolution, imgs_ds.shape[1]))
+        # check energy conservation after downsample
+        self.assertTrue(
+            np.allclose(imgs_org.norm(), self.L / max_resolution * imgs_ds.norm())
+        )
+
+    def testNormBackground(self):
+        bg_radius = 1.0
+        grid = grid_2d(self.L)
+        mask = grid["r"] > bg_radius
+        self.sim.normalize_background()
+        imgs_nb = self.sim.images(start=0, num=self.n).asnumpy()
+        new_mean = np.mean(imgs_nb[:, mask])
+        new_variance = np.var(imgs_nb[:, mask])
+
+        # new mean of noise should be close to zero and variance should be close to 1
+        self.assertTrue(new_mean < 1e-7 and abs(new_variance - 1) < 1e-7)
+
+    def testWhiten(self):
+        noise_estimator = WhiteNoiseEstimator(self.sim)
+        self.sim.whiten(noise_estimator.filter)
```

Comment:
Noise in simulation is already white, so this has no effect. Change it to some other noise type.

Reply:
Changed a different filter in commit 999145c.

Reply:
Ok. Why is it a `PowerFilter` with `power=0.2`? Isn't it enough with just the `FunctionFilter`?

Reply:
It depends which value to use as criterion. If I just use the PowerFilter, correlation coefficient is 0.52755233. Can the whiten do a good job for noise far away Gassian?  The power = 0.2  make correlation coefficient smaller than 0.1.

Reply:
I don't understand. It's a consistent estimator of the power spectrum so given enough samples, it will converge to the proper one. Why do you want a small correlation coefficient?

That being said, you may run into problems if the power spectrum decays to zero. In that case, you'll be able to estimate the power spectrum ok, but whitening will be difficult since you effectively divide by zero. You may be better off with some 1/f-type noise in that case.

Reply:
It looks that I need to use `AnisotropicNoiseEstimator` for estimating the noise when I use the FunctionFilter directly, in order to have points with low correlation coefficients. Changed in new commit e6bb24a